node-graph/gstd/src/wasm_application_io.rs

```diff
@@ -288,77 +289,129 @@ fn decode_image_node<'a: 'input>(data: Arc<[u8]>) -> ImageFrame<Color> {
+	}
+	/*
+	let preamble = "data:image/svg+xml;base64,";
+	let mut base64_string = String::with_capacity(preamble.len() + array.len() * 4);
+	base64_string.push_str(preamble);
+	base64::engine::general_purpose::STANDARD.encode_string(array, &mut base64_string);
+
+	let image_data = web_sys::HtmlImageElement::new().unwrap();
+	image_data.set_src(base64_string.as_str());
+	wasm_bindgen_futures::JsFuture::from(image_data.decode()).await.unwrap();
+	context.draw_image_with_html_image_element(&image_data, 0.0, 0.0).unwrap();
+	*/
+	let frame = SurfaceHandleFrame {
+		surface_handle,
+		transform: DAffine2::IDENTITY,
+	};
+	RenderOutput::CanvasFrame(frame.into())
+}
 
-	match output_format {
-		ExportFormat::Svg => {
-			data.render_svg(&mut render, &render_params);
-			// TODO: reenable once we switch to full node graph
-			let min = footprint.transform.inverse().transform_point2((0., 0.).into());
-			let max = footprint.transform.inverse().transform_point2(resolution.as_dvec2());
-			render.format_svg(min, max);
-			RenderOutput::Svg(render.svg.to_string())
-		}
-		#[cfg(any(feature = "resvg", feature = "vello"))]
-		ExportFormat::Canvas => {
-			data.render_svg(&mut render, &render_params);
-			// TODO: reenable once we switch to full node graph
-			let min = footprint.transform.inverse().transform_point2((0., 0.).into());
-			let max = footprint.transform.inverse().transform_point2(resolution.as_dvec2());
-			render.format_svg(min, max);
-			let string = render.svg.to_string();
-			let array = string.as_bytes();
-			let canvas = &surface_handle.surface;
-			canvas.set_width(resolution.x);
-			canvas.set_height(resolution.y);
-			let usvg_tree = data.to_usvg_tree(resolution, [min, max]);
-
-			if let Some(exec) = editor.application_io.gpu_executor() {
-				todo!()
-			} else {
-				let rtree = resvg::Tree::from_usvg(&usvg_tree);
-
-				let pixmap_size = rtree.size.to_int_size();
-				let mut pixmap = resvg::tiny_skia::Pixmap::new(pixmap_size.width(), pixmap_size.height()).unwrap();
-				rtree.render(resvg::tiny_skia::Transform::default(), &mut pixmap.as_mut());
-				let array: Clamped<&[u8]> = Clamped(pixmap.data());
-				let context = canvas.get_context("2d").unwrap().unwrap().dyn_into::<CanvasRenderingContext2d>().unwrap();
-				let image_data = web_sys::ImageData::new_with_u8_clamped_array_and_sh(array, pixmap_size.width(), pixmap_size.height()).expect("Failed to construct ImageData");
-				context.put_image_data(&image_data, 0.0, 0.0).unwrap();
+// Render with the data node taking in Footprint.
+impl<'input, 'a: 'input, T: 'input + GraphicElementRendered, F: 'input + Future<Output = T>, Data: 'input, Surface: 'input, SurfaceFuture: 'input> Node<'input, WasmEditorApi<'a>>
+	for RenderNode<Data, Surface, Footprint>
+where
+	Data: Node<'input, Footprint, Output = F>,
+	Surface: Node<'input, (), Output = SurfaceFuture>,
+	SurfaceFuture: core::future::Future<Output = Arc<SurfaceHandle<HtmlCanvasElement>>>,
```

Comment:
What is the reason for no longer using the macro here?

Reply:
It doesn't work with the third generic argument (used to allow the upstream node to take either a `Footprint` or `()`)

Reply:
I guess I am still confused why we need this, when do we not have a Footprint?

Reply:
If the user creates a rectangle node, then a fill node, then connects that to the output.

Reply:
![rectangle -> fill -> output graph](https://github.com/GraphiteEditor/Graphite/assets/78500760/0460a911-54d5-4ff7-a442-c8755e93cd59)

Reply:
Ah, so the issue is the missing `CullNode` ?

Reply:
Also if the user wishes to preview a colour node:
![Previewing a colour node](https://github.com/GraphiteEditor/Graphite/assets/78500760/b790e36a-5324-4e07-b02e-46a4756f6c2f)

Reply:
In both cases I would prefer if we still used the IntoNode to do the conversion for us instead of making the render node overly generic to improve code reuse

Reply:
Could you elaborate on what you mean by using the into node? What is converted into what? Who inserts the into node?

Reply:
![image](https://github.com/GraphiteEditor/Graphite/assets/21245806/e9c4f86a-f537-4ee0-9dbd-8ab5e55f27a9)
basically this thing, this does the conversion and the render node only gets a proper output

Reply:
This `IntoNode` still requires a node that has a `Footprint` for input. We want to render nodes that have `()` for input.

We also do not want to convert to a graphic group. `bool`s can't be converted into a graphic group, yet they should still be previewed.

Reply:
what should the preview for a bool look like? Also does it make sense to connect it to the render node at all in that case?

Reply:
If we don't connect the data to the render node, the graphite editor can decide how to display the data

Reply:
![boolean preview](https://github.com/GraphiteEditor/Graphite/assets/78500760/2ef7f666-0381-4a3e-b928-3d6034822a1a)
It makes sense to connect to the boolean node to the render node because it is rendered into the document.

Reply:
> If we don't connect the data to the render node, the graphite editor can decide how to display the data

I'm not sure I understand the advantage of having the RenderNode render only some types of data, and the editor render other types of data. Would it not be easier to render everything in one place?

Reply:
Ideally I'd like the render node to be as lightweight as possible and move as much of the functionality out of it as possible to make it reusable and composable. I don't really think about displaying bools as "rendering" per se, that feels more like a debug output. If the user were to for example connect a "spreadsheet" to the output I think it would make more sense to show a spreadsheet UI from which the user can copy paste data, search for values etc. rather than rendering that to an SVG.

Reply:
How about wrapping the render node, which produces only graphical content, with another node which also includes "debug"-style non-graphical data type visualizations? That might help to keep it lightweight as you'd desire while still preserving this PR's cool behavior as it's currently implemented.

From a product design perspective, it remains an open question what the desired final behavior should look like for users viewing non-graphical data. But this current approach is quite a decent stand-in that provides useful value to the user in absence of any other fancy UIs like the spreadsheet panel you mentioned. Having something at all serves as a good angle to imagine what further improvements could look like. So I'd definitely like the current behavior to be merged, although I'll leave the code structuring up to the two of you as you're both more knowledgeable than me about the code.

Reply:
But to reiterate my point about the non-use of the macro, shouldn't it be possible to describe one of the use cases using the `node_fn` macro and then add the second implementation using the `node_impl` macro?

Reply:
I guess the way I would go about this is to just reuse the monitor nodes for this and by clicking the monitor node we could allow the user to preview the data that flows through it.  This would allow us to not have to channel everything through the render node which would be an architecturally much cleaner solution. (with the current solution this would mess up caching etc.)

Reply:
> This leads to an error as the PhantomData is not properly handled.

Yeah, I've since tried that out as well although I have to say that I'm confused as to why the implementations are in conflict with each other without using the parameter phantom data, could be that the rust type system is not powerful enough or that I misunderstand something about the setup

Reply:
> I guess the way I would go about this is to just reuse the monitor nodes for this

This wouldn't work for previewing a node that the user just placed into the graph without connecting to the output (e.g. both examples above) as the monitor node would never be run.

Reply:
I have made it so the executor renders things in the case of a node being previewed. I have kept the current behaviour in the render node so it works if the user decides to connect their bool to the output node (which they might do if they didn't see the preview button).

Reply:
Perhaps? So when doing typing for the protonodes, if it found a node with no matching implementation, it would check to see if the given input is a Footprint and if there is an implementation with input of `()`?